lib/resources/postgres.rb

```diff
@@ -18,48 +19,26 @@ def initialize
         # Debian allows multiple versions of postgresql to be
         # installed as well as multiple "clusters" to be configured.
         #
-        version = version_from_dir('/etc/postgresql')
-        cluster = cluster_from_dir("/etc/postgresql/#{version}")
-        @conf_dir = "/etc/postgresql/#{version}/#{cluster}"
-        @data_dir = "/var/lib/postgresql/#{version}/#{cluster}"
-      elsif os.redhat?
-        #
-        # /var/lib/pgsql/data is the default data directory on RHEL6
-        # and RHEL7. However, PR #824 explicitly added version-based
-        # directories. Thus, we call #version_from_dir unless it looks
-        # like we are using unversioned directories.
-        #
-        # TODO(ssd): This has the potential to be noisy because of the
-        # warning in version_from_dir. We should determine which case
-        # is more common and only warn in the less common case.
-        #
-        version = if inspec.directory('/var/lib/pgsql/data').exist?
-                    warn 'Found /var/lib/pgsql/data. Assuming postgresql install uses un-versioned directories.'
-                    nil
-                  else
-                    version_from_dir('/var/lib/pgsql/')
-                  end
-
-        @data_dir = File.join('/var/lib/pgsql/', version.to_s, 'data')
-      elsif os[:name] == 'arch'
-        #
-        # https://wiki.archlinux.org/index.php/PostgreSQL
-        #
-        # The archlinux wiki points to /var/lib/postgresql/data as the
-        # main data directory.
-        #
-        @data_dir = '/var/lib/postgres/data'
+        @version = version_from_psql || version_from_dir('/etc/postgresql')
+        @cluster = cluster_from_dir("/etc/postgresql/#{@version}")
+        @conf_dir = "/etc/postgresql/#{@version}/#{@cluster}"
+        @data_dir = "/var/lib/postgresql/#{@version}/#{@cluster}"
       else
-        #
-        # According to https://www.postgresql.org/docs/9.5/static/creating-cluster.html
-        #
-        # > There is no default, although locations such as
-        # > /usr/local/pgsql/data or /var/lib/pgsql/data are popular.
-        #
-        @data_dir = '/var/lib/pgsql/data'
+        @version = version_from_psql
+        if @version.nil?
+          if inspec.directory('/var/lib/pgsql/data').exist?
+            warn 'Unable to determine PostgreSQL version: psql did not return
+             a version number and unversioned data directories were found.'
+            nil
+          else
+            version_from_dir('/var/lib/pgsql/')
+          end
+        end
+        verify_data_dir_location(@version)
```

Comment:
I'd like to see us remain consistent within a resource whether helper functions manipulate instance variables, or whether they return data to the caller who then decides whether to update an instance variable.

I prefer that methods do NOT manipulate instance variables because it makes it harder to test them and also could unknowingly pollute other tests.

I'd like to see this be:

```
@data_dir = verify_data_dir_locaton(@version)
```

... and have `#verify_data_dir_location` changed to `#locate_data_dir_location_by_version` and have it return a string containing the validated data directory.